core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/variable/listener/support/NotifiableRegistry.java

```diff
@@ -0,0 +1,97 @@
+            new LinkedHashMap<>();
+    private final Map<VariableDescriptor<?>, List<VariableListenerNotifiable<Solution_>>> sourceVariableToNotifiableMap =
+            new LinkedHashMap<>();
+    private int nextGlobalOrder = 0;
+
+    NotifiableRegistry(InnerScoreDirector<Solution_, ?> scoreDirector) {
+        this.scoreDirector = scoreDirector;
+        for (EntityDescriptor<Solution_> entityDescriptor : scoreDirector.getSolutionDescriptor().getEntityDescriptors()) {
+            sourceEntityToNotifiableMap.put(entityDescriptor, new LinkedHashSet<>());
+            for (VariableDescriptor<Solution_> variableDescriptor : entityDescriptor.getDeclaredVariableDescriptors()) {
+                sourceVariableToNotifiableMap.put(variableDescriptor, new ArrayList<>());
+            }
+        }
+    }
+
+    void registerShadowVariableListener(ShadowVariableDescriptor<Solution_> shadowVariableDescriptor,
+            VariableListener<Solution_, ?> variableListener) {
+        int globalOrder = shadowVariableDescriptor.getGlobalShadowOrder();
+        VariableListenerNotifiable<Solution_> notifiable =
+                new VariableListenerNotifiable<>(scoreDirector, variableListener, globalOrder);
+        for (VariableDescriptor<Solution_> source : shadowVariableDescriptor.getSourceVariableDescriptorList()) {
+            registerNotifiable(source, notifiable);
+        }
+        notifiableList.add(notifiable);
+        nextGlobalOrder = Math.max(nextGlobalOrder, globalOrder + 1);
+    }
+
+    void registerSourcedVariableListener(SourcedVariableListener<Solution_, ?> variableListener) {
+        VariableListenerNotifiable<Solution_> notifiable =
+                new VariableListenerNotifiable<>(scoreDirector, variableListener, nextGlobalOrder);
+        VariableDescriptor<Solution_> source = variableListener.getSourceVariableDescriptor();
+        registerNotifiable(source, notifiable);
+        notifiableList.add(notifiable);
+        nextGlobalOrder++;
+    }
+
+    private void registerNotifiable(VariableDescriptor<?> source, VariableListenerNotifiable<Solution_> notifiable) {
+        sourceVariableToNotifiableMap.get(source).add(notifiable);
+        sourceEntityToNotifiableMap.get(source.getEntityDescriptor()).add(notifiable);
+    }
+
+    void sort() {
+        Collections.sort(notifiableList);
+    }
+
+    Iterable<VariableListenerNotifiable<Solution_>> getAll() {
+        return notifiableList;
+    }
+
+    Collection<VariableListenerNotifiable<Solution_>> get(EntityDescriptor<?> entityDescriptor) {
+        return sourceEntityToNotifiableMap.get(entityDescriptor);
+    }
+
+    Collection<VariableListenerNotifiable<Solution_>> get(VariableDescriptor<?> variableDescriptor) {
+        return sourceVariableToNotifiableMap.getOrDefault(variableDescriptor,
+                Collections.emptyList()); // Avoids null for chained swap move on an unchained var.
+    }
```

Comment:
I am not too excited that we are returning internal data structures to the outside code.

This allows callers to modify internal state of this class from the outside, and that in turn makes it hard to figure out who has access to what. I know that OP core uses this pattern religiously, but perhaps we can start changing that?

Reply:
Generally, I would agree but in this case I can't see the benefit of hiding the collection. Given how it's being used on the caller side, using something else (e.g. an Iterator) would make the code awkward IMO. Furthermore, this a package-private class, so figuring out the usages is not difficult.